pkg/operation/botanist/controlplane/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,553 @@
+}
+
+func (r *ResourceManager) ensureRoleBinding(ctx context.Context) error {
+	roleBinding, err := r.emptyRoleBindingInWatchedNamespace()
+	if err != nil {
+		return err
+	}
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, roleBinding, func() error {
+		roleBinding.Labels = *r.config.DefaultLabels
+		roleBinding.RoleRef = rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     roleName,
+		}
+		roleBinding.Subjects = []rbacv1.Subject{{
+			Kind:      rbacv1.ServiceAccountKind,
+			Name:      serviceAccountName,
+			Namespace: r.namespace,
+		}}
+		return nil
+	})
+	return err
+}
+
+func (r *ResourceManager) emptyRoleBindingInWatchedNamespace() (*rbacv1.RoleBinding, error) {
+	if r.config.WatchedNamespace == nil {
+		return nil, errors.New("creating RoleBinding in watched namespace failed - no namespace defined")
+	}
+	return &rbacv1.RoleBinding{ObjectMeta: metav1.ObjectMeta{Name: "gardener-resource-manager", Namespace: *r.config.WatchedNamespace}}, nil
+}
+
+func (r *ResourceManager) ensureService(ctx context.Context) error {
+	const (
+		healthPortName  = "health"
+		metricsPortName = "metrics"
+	)
+
+	service := r.emptyService()
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, service, func() error {
+		service.Labels = *r.config.DefaultLabels
+		service.Spec.Selector = *appLabel()
+		service.Spec.Type = corev1.ServiceTypeClusterIP
+		service.Spec.ClusterIP = corev1.ClusterIPNone
+		service.Spec.Ports = kutil.ReconcileServicePorts(service.Spec.Ports, []corev1.ServicePort{
+			{
+				Name:     metricsPortName,
+				Protocol: corev1.ProtocolTCP,
+				Port:     metricsPort,
+			},
+			{
+				Name:     healthPortName,
+				Protocol: corev1.ProtocolTCP,
+				Port:     healthPort,
+			},
+		})
+		return nil
+	})
+	return err
+}
+
+func (r *ResourceManager) emptyService() *corev1.Service {
+	return &corev1.Service{ObjectMeta: metav1.ObjectMeta{Name: "gardener-resource-manager", Namespace: r.namespace}}
+}
+
+func (r *ResourceManager) ensureDeployment(ctx context.Context) error {
+	const (
+		limitCPU                  = "400m"
+		limitMemory               = "512Mi"
+		requestCPU                = "23m"
+		requestMemory             = "47Mi"
+		volumeMountName           = "gardener-resource-manager"
+		volumeMountPathKubeconfig = "/etc/gardener-resource-manager"
+		volumeName                = "gardener-resource-manager"
+	)
+
+	deployment := r.emptyDeployment()
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, deployment, func() error {
+		deployment.Labels = *r.config.DefaultLabels
+
+		deployment.Spec.Replicas = &r.replicas
+		deployment.Spec.RevisionHistoryLimit = pointer.Int32Ptr(0)
+		deployment.Spec.Selector = &metav1.LabelSelector{MatchLabels: *appLabel()}
+
+		deployment.Spec.Template = corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: utils.MergeStringMaps(*r.config.DefaultLabels, map[string]string{
```

Comment:
Where is the app label?

Reply:
DefaultLabels (now renamed to Labels) always includes the app label. This is ensured in the constructor : https://github.com/gardener/gardener/pull/3445/files#diff-a66ff079fa067226cf9c172079a286e54f39e2a5ec6ad863136488759de35274R90-R95

Reply:
Hm, OK, I missed that, it's a little confusing. Can we remove the `Labels` attribute from the config values and rather compute it in the package itself, similar to how it's done in all other component packages?

Reply:
OK, got it. That would mean I would parse if a target kubeconfig is configured and then in this case add the `v1beta1constants.GardenRoleControlPlane` label ? Something like this?
I feel this information should be passed via the constructor because the caller actually knows this and not the resource manager.

Reply:
Well, no strong opinion. I can only refer to what I wrote in https://github.com/gardener/gardener/pull/3445#discussion_r566736098. With the "similarity" argument I'd prefer to remove it from here (because, as you can see, I was already confused because this package handles some things different than other packages), but it's a detail with low value, so I'm fine with whatever you decide.